File: codes/list/142.linked-list-cycle-ii/linked-list-cycle-ii.go
package main

type ListNode struct {
    Val int
    Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	fast:=head
	slow:=head
	for fast!=nil&&fast.Next!=nil{
		fast=fast.Next.Next
		slow=slow.Next
		if slow==fast{
			goto CYCLED
		}
	}
	return nil

	CYCLED:
	slow=head
	for slow!=fast{
		slow=slow.Next
		fast=fast.Next
	}
	return slow
}

File: codes/list/142.linked-list-cycle-ii/linked-list-cycle-ii.cpp
#include "ListNode.h"
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        auto slow = head, fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                goto CYCLED;
            }
        }
        return nullptr;
    CYCLED:
        slow = head;
        while (slow != fast) {
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
    }
};